csharp/src/Ice/InputStream.cs

```diff
@@ -217,6 +262,87 @@ public string ReadString()
             return value;
         }
 
+        /// <summary>Reads a uint from the stream.</summary>
+        /// <returns>The uint read from the stream.</returns>
+        public uint ReadUInt()
+        {
+            uint value = BitConverter.ToUInt32(_buffer.AsSpan(_pos, sizeof(uint)));
+            _pos += sizeof(uint);
+            return value;
+        }
+
+        /// <summary>Reads a ulong from the stream.</summary>
+        /// <returns>The ulong read from the stream.</returns>
+        public ulong ReadULong()
+        {
+            ulong value = BitConverter.ToUInt64(_buffer.AsSpan(_pos, sizeof(ulong)));
+            _pos += sizeof(ulong);
+            return value;
+        }
+
+        /// <summary>Reads a ushort from the stream.</summary>
+        /// <returns>The ushort read from the stream.</returns>
+        public ushort ReadUShort()
+        {
+            ushort value = BitConverter.ToUInt16(_buffer.AsSpan(_pos, sizeof(ushort)));
+            _pos += sizeof(ushort);
+            return value;
+        }
+
+        /// <summary>Reads an int from the stream. This int is encoded using Ice's variable-length integer encoding.
+        /// </summary>
+        /// <returns>The int read from the stream.</returns>
+        public int ReadVarInt()
+        {
+            try
+            {
+                return (int)ReadVarLong();
+            }
+            catch (Exception ex)
+            {
+                throw new InvalidDataException("varint value is out of range", ex);
+            }
+        }
+
+        /// <summary>Reads a long from the stream. This long is encoded using Ice's variable-length integer encoding.
+        /// </summary>
+        /// <returns>The long read from the stream.</returns>
+        public long ReadVarLong() => OldEncoding ? ReadLong() :
+            (_buffer[_pos] & 0x03) switch
+            {
+                0 => (sbyte)ReadByte() >> 2,
+                1 => ReadShort() >> 2,
+                2 => ReadInt() >> 2,
+                _ => ReadLong() >> 2
+            };
+
+        /// <summary>Reads a uint from the stream. This uint is encoded using Ice's variable-length integer encoding.
+        /// </summary>
+        /// <returns>The uint read from the stream.</returns>
+        public uint ReadVarUInt()
+        {
+            try
+            {
+                return (uint)ReadVarULong();
+            }
+            catch (Exception ex)
+            {
+                throw new InvalidDataException("varuint value is out of range", ex);
+            }
+        }
+
+        /// <summary>Reads a ulong from the stream. This ulong is encoded using Ice's variable-length integer encoding.
+        /// </summary>
+        /// <returns>The ulong read from the stream.</returns>
+        public ulong ReadVarULong() => OldEncoding ? ReadULong() :
+            (_buffer[_pos] & 0x03) switch
+            {
+                0 => (uint)ReadByte() >> 2,
+                1 => (uint)ReadUShort() >> 2,
```

Comment:
It's not clear why we need the cast to uint here, perhaps add a comment?

Reply:
Added comment.